File: 0368-largest-divisible-subset/0368-largest-divisible-subset.java
class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        Arrays.sort(nums);
        int n = nums.length;
        int dp[] = new int[n];
        Arrays.fill(dp,1);
        int hash[] = new int[n];

        int max =1 , lastIndex = 0;
        for(int i=1;i<n;i++){
            hash[i] = i;
            for(int j=0;j<i;j++){
                if(nums[i] %nums[j] == 0 && dp[j]+1 > dp[i]){
                    dp[i] = dp[j]+1;
                    hash[i] = j;
                }
            }
            if(dp[i] > max){
                max = dp[i];
                lastIndex = i;
            }
        }
        List<Integer> list = new ArrayList<>();
        
        list.add(nums[lastIndex]);

        while(lastIndex != hash[lastIndex]){
            lastIndex = hash[lastIndex];
            list.add(nums[lastIndex]);
        }
        Collections.reverse(list);
        return list;
    }
}

File: 0368-largest-divisible-subset/0368-largest-divisible-subset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n,1) , hash(n,1);
        sort(nums.begin(),nums.end());
        int maxi = 1;
        int idx =-1;
        for(int i=0;i<n;i++){
            hash[i] = i;
            for(int prev=0;prev<i;prev++){
                if(nums[i]%nums[prev] == 0 && dp[prev]+1 > dp[i] ){
                    dp[i] = dp[prev]+1;
                    hash[i]= prev;
                }
            }
            if(idx == -1 || dp[i]>maxi){
                maxi = dp[i];
                idx = i;
            } 
        }
        vector<int> temp;
        temp.push_back(nums[idx]);
        while(hash[idx] != idx){
            idx =  hash[idx];
            temp.push_back(nums[idx]);
        }
        reverse(temp.begin(),temp.end());
        return temp;
    }
};